### PR TITLE
Fix many issues with make_perturbed_fun

### PR DESCRIPTION
**Pull request by adconner**
Issues fixed:
1) Handle arbitrarily shaped arrays in inputs.
   See #1309 
2) Scale the jvp correctly. The previous implementation was missing a
   factor of 1/sigma (see Prop 3.1 of https://arxiv.org/abs/2002.08676
   for correct formula).
   See #1308 

Enhancements:
1) Vmap over samples instead of using unrolled loops. This greatly
   improves the previously terrible compile times and should also scale
   better on hardware.
2) Use custom_jvp.defjvp instead of custom_jvp.defjvps. This allows us
   to be explicit that the rng input is not differentiable, and also
   avoids calculating _compute_residuals twice during jvp.
3) Use jax.jvp instead of jax.grad for the computation of the
   grad log_prob inner product. This is minor, but potentially valuable
   for more complicated probability distributions.

I also considered making the design change discussed in #1214, where it was proposed to use the interface sample_and_log_prob, but I now believe this is impossible. We don't actually care about the log_prob for our samples, but its gradient. The data dependence through the differentiation of log_prob applied to the result of sample requires two separate forward passes, even if we have an efficient function which can compute the log_prob value along with sample in only one pass.

### CODE REVIEW

**Comment by rdyro** [comment]
This is a really useful contribution, thanks!

Google requires you to sign the CLA first and I should be able to review your changes soon, I'd love to help merge this to also resolve the issues you opened.

**Comment by adconner** [comment]
CLA signed. 

Looks like linter complains that test loss functions are defined in a loop in the test runner. I could replace them with a less readable one line lambda or lift them out of the loop with an extra argument and use functools.partial to bind it later. Neither seems like much of a code improvement to me. Are either of these changes desired?

**Comment by rdyro** [comment]
It shouldn't change anything in your code, but this "loop variable close over" sometimes leads to pretty insidious bugs in python if you're not immediately consuming your closures (which you are).

It's a little annoying, but if you could refactor, it'd be great.

**Comment by rdyro** [comment]
I apologize this linter feedback loop is long, I'm trying to add a local check to pre-commit here: [pull/1313](https://github.com/google-deepmind/optax/pull/1313). For now if you want to check it locally
```
ruff check --select=B023 {DIRECTORY}
```

**Comment by adconner** [comment]
Sorry for all the pings. In hindsight I should have just looked at what the github actions were to see which lints were being run! 

In the end I also had to loosen up the test tolerances as they did not pass with high enough probability.

File: optax/perturbations/_make_pert.py [hunk 1]
```diff
@@ -130,32 +101,31 @@ def _compute_residuals(
       inputs: chex.ArrayTree, rng: chex.PRNGKey
   ) -> tuple[chex.ArrayTree, chex.ArrayTree]:
     # random noise Zs to be added to inputs
-    samples = [
-        optax.tree.random_like(rng_, inputs, sampler=noise.sample)
-        for rng_ in jax.random.split(rng, num_samples)
-    ]
+    samples = jax.vmap(lambda rng_: optax.tree.random_like(rng_, inputs, sampler=noise.sample))(
```
Comment (rdyro) [inline optax/perturbations/_make_pert.py]: Can we use:     
```python
    broadcast_shapes = jax.tree.map(
        lambda x: jax.ShapeDtypeStruct((num_samples,) + x.shape, x.dtype),
        inputs
    )
    samples = optax.tree.random_like(rng, broadcast_shapes, sampler=noise.sample)
```
to save splitting the random key `num_samples` of times.

File: optax/perturbations/_make_pert_test.py [hunk 2]
```diff
@@ -81,40 +87,53 @@ def setUp(self):
...
     ]
 
-  def test_pert_close_array(self):
+    # linear function x -> Ax + b
+    self.linear_A = jnp.array([[0.0, 1.0, 2.0],
+                               [3.0, 4.0, 5.0]])
+    self.linear_b = jnp.array([1.0,-2.0])
+    self.linear_x_test = jnp.array([-1.0, 1.0, 0.0])
+
+  def test_pert_argmax(self):
```
Comment (rdyro) [inline optax/perturbations/_make_pert_test.py]: Prefer passing the key to the test explicitly and decorating the test like so (the old design was bad):
```python
  @parameterized.product(key=[0, 1, 2])
  def test_pert_argmax(self, key):
    """Test that make_perturbed_fun matches theoretical expression of gradient.

    Applies to the case of an argmax. Includes a test for the gradients and the
    Hessian of a scalar loss.
    """
    
    rng_jax = jax.random.key(key)
```
Applies to all tests (fixing this would be great!)

File: optax/perturbations/_make_pert_test.py [hunk 3]
```diff
@@ -45,17 +47,21 @@ def setUp(self):
     rng = np.random.RandomState(0)
 
     self.rng_jax = jax.random.key(0)
```
Comment (rdyro) [inline optax/perturbations/_make_pert_test.py]: Could we change this test class to not use a persistent random key, this doesn't follow the practice in the rest of the repo. We didn't catch this before.

Could you delete this line and use `parameterized.product` to pass keys to each test, see other comment.

File: optax/perturbations/_make_pert.py (hunk 1 above)
Comment (carlosgmartin) [reply optax/perturbations/_make_pert.py]: @rdyro [`tree_batch_shape`](https://optax.readthedocs.io/en/latest/api/utilities.html#optax.tree_utils.tree_batch_shape) can be used for this.

File: optax/perturbations/_make_pert_test.py (hunk 2 above)
Comment (carlosgmartin) [reply optax/perturbations/_make_pert_test.py]: @rdyro Shouldn't that argument be called `seed`?

Comment (rdyro) [reply optax/perturbations/_make_pert_test.py]: That's right, it should be `seed`!

```python
  @parameterized.product(seed=[0, 1, 2])
  def test_pert_argmax(self, seed):
    """Test that make_perturbed_fun matches theoretical expression of gradient.

    Applies to the case of an argmax. Includes a test for the gradients and the
    Hessian of a scalar loss.
    """
    
    key = jax.random.key(seed)
```

Thanks @carlosgmartin !

**Comment by vroulet** [comment]
Great PR! Thank you @adconner ! 
The function is not twice differentiable not because jax would not be able to actually differentiate through the jvp but because it would give the wrong estimator of the second derivative right? (cf https://arxiv.org/abs/1802.05098)
If you agree, we can add this rather than just saying that the function is not twice differentiable. (I haven't actually tried to compute a second derivative, maybe jax would actually throw an error).

**Comment by vroulet** [comment]
> I also considered making the design change discussed in https://github.com/google-deepmind/optax/issues/1214, where it was proposed to use the interface sample_and_log_prob, but I now believe this is impossible. We don't actually care about the log_prob for our samples, but its gradient. The data dependence through the differentiation of log_prob applied to the result of sample requires two separate forward passes, even if we have an efficient function which can compute the log_prob value along with sample in only one pass.

We could (in another PR if you want) consider removing the classes and add clear arguments for sampler and log_prob.

**Comment by carlosgmartin** [comment]
Can we add a test that would've caught the incorrectness of the previous implementation, and checks the correctness of this one?

Perhaps something along the following lines:

```python3
import jax
import optax
from jax import numpy as jnp
from jax import random


def simple_make_perturbed_fun(
    f,
    num_samples=1000,
    sigma=0.1,
    noise=optax.perturbations.Gumbel(),
):
    # Monte Carlo estimate of E[f(x + σ Z)] where Z ~ noise
    # only works for differentiable f
    def g(x, key):
        zs_shape = optax.tree_utils.tree_batch_shape(x, [num_samples])
        zs = optax.tree_utils.tree_random_like(key, zs_shape, noise.sample)
        xs = optax.tree_utils.tree_add_scale(x, sigma, zs)
        ys = jax.vmap(f)(xs)
        ys_mean = jax.tree.map(lambda leaf: jnp.mean(leaf, 0), ys)
        return ys_mean

    return g


def test_make_perturbed_fun(
    num_samples=1000,
    sigma=1,
    noise=optax.perturbations.Normal(),
    atol=1e-8,
    seed=0,
):
    def f(x):
        # a differentiable test function
        x0, x1, x2 = x
        y0 = x0 * x1**2 * jnp.sin(x2 * x1 - x0) + jnp.cos(x0**3)
        y1 = x2 + x0 * jnp.sin(x1)
        return jnp.stack([y0, y1])

    f1 = optax.perturbations.make_perturbed_fun(
        f,
        num_samples=num_samples,
        sigma=sigma,
        noise=noise,
    )
    f2 = simple_make_perturbed_fun(
        f,
        num_samples=num_samples,
        sigma=sigma,
        noise=noise,
    )

    x = jnp.array([0.3, 0.4, 0.5])
    key = random.key(seed)
    j1 = jax.jacobian(f1)(x, key)
    j2 = jax.jacobian(f2)(x, key)
    print(j1)
    print(j2)
    assert jnp.allclose(j1, j2, atol=atol)


if __name__ == "__main__":
    test_make_perturbed_fun()
```

**Comment by rdyro** [comment]
What is this test testing @carlosgmartin ? Is a check against a simpler reference implementation?

**Comment by carlosgmartin** [comment]
@rdyro Yes (but the simpler version only works for differentiable functions; so we run both on a differentiable function and compare).

**Comment by rdyro** [comment]
> @rdyro Yes (but the simpler version only works for differentiable functions; so we run both on a differentiable function and compare).

Let's add it in an immediate follow up PR since this one has quite a few moving parts already. @adconner would you be able to contribute the additional test in a follow up PR? @carlosgmartin or if you wanted to do it?

**Comment by adconner** [comment]
> Great PR! Thank you @adconner ! The function is not twice differentiable not because jax would not be able to actually differentiate through the jvp but because it would give the wrong estimator of the second derivative right? (cf https://arxiv.org/abs/1802.05098) If you agree, we can add this rather than just saying that the function is not twice differentiable. (I haven't actually tried to compute a second derivative, maybe jax would actually throw an error).

Differentiating through the jvp would just differentiate through the Monte-Carlo estimation, rather than express the result as a new Monte-Carlo estimation. Put more directly, differentiating through the perturbed jvp differentiates the original function, which we are trying to avoid. 

I think it may be possible to make this work though, one just needs to express the jvp as another perturbed function. It looks like maybe one can take f perturbed with respect to a modified distribution. I need to think a bit more about it.

This suggests a test should be added testing that the derivative of f is not used in the derivative of f_perturb (by for example wrapping f with a custom jvp which errors out and calling the jvp of f_peturb).

> We could (in another PR if you want) consider removing the classes and add clear arguments for sampler and log_prob.

I'm happy with whatever you want to do here. This change is relatively orthogonal to the present one if we agree that we still want to use sampler and log_prob separately. Although the current interface is pretty ergonomic with the distributions in distrax, eg, as distributions are specified as classes providing at least the functions `sample` and `log_prob`

> Let's add it in an immediate follow up PR since this one has quite a few moving parts already. @adconner would you be able to contribute the additional test in a follow up PR? @carlosgmartin or if you wanted to do it?

I can do it if you like.

**Comment by rdyro** [comment]
I appreciate the comments! To summarize:

@adconner could you:
* add your explanation to the docstring for why the higher order derivatives are non-informative / not supported in the current implementation? let's not handle the 2nd order derivative behavior for now (in this PR), but plan to either support it or catch it and error
* include @carlosgmartin 's additional test
* defer "removing the classes and add clear arguments for sampler and log_prob" to another PR

**Review by rdyro — APPROVED** [review]
Great PR, thanks so much!

**Comment by vroulet** [comment]
@adconner 

> I think it may be possible to make this work though, one just needs to express the jvp as another perturbed function. It looks like maybe one can take f perturbed with respect to a modified distribution. I need to think a bit more about it.

This is what the dice paper is about https://arxiv.org/abs/1802.05098

There is some further literature on "automatic stochastic estimates of derivatives" like this [paper](https://proceedings.neurips.cc/paper_files/paper/2021/file/3dfe2f633108d604df160cd1b01710db-Paper.pdf) or even [this one](https://arxiv.org/pdf/2212.06386).

Essentially the make_perturb implements a score function estimator (aka REINFORCE). To enable higher automatic differentiation one should use the "magic_box" operator of the dice paper I think.
